### ISSUE TITLE
[caffe2] segmentation fault-  workspace.RunNet(model)

### ISSUE THREAD

**Issue by dollysapra** [pytorch/pytorch#17616]
I am new to GPU based training, I recently moved my code to a system with CUDA-10, my code which is working on CPU only mode is throwing segmentation faults at runtime on trying to run workspace.runnet(model) with GPU support. 

I am using-  with core.DeviceOption(Caffe2_pb2.CUDA,0):  while building my model. no other change. Do i need to do something else? 

Appreciate your help.

**Comment by dollysapra** [comment]
Hello,
Instead of using with core.deviceScope, I build the model normally and used the helper function RunAllOnGPU - 
```
    train_model.param_init_net.RunAllOnGPU() 
    train_model.net.RunAllOnGPU()
```

Now I donot get segmentation fault but an error - RuntimeError: [enforce fail at momentum_sgd_op.h:90] OperatorBase::InputIsTensorType(GRAD, device_type). 
Error from operator: 
input: "conv303weights_grad" input: "conv303weights_momentum" input: "SgdOptimizer_0_lr_cpu" input: "conv303weights" output: "conv303weights_grad" output: "conv303weights_momentum" output: "conv303weights" name: "" type: "MomentumSGDUpdate" arg { name: "momentum" f: 0.9 } arg { name: "nesterov" i: 1 } device_option { }

Here I notice SgdOptimizer_0_lr_cpu - should cpu be there in the auto-generated name? Any thing else I should be doing differntly? 
```
   optimizer.build_sgd(train_model,base_learning_rate= 0.02,
            policy="fixed",momentum=0.9,
            weight_decay=0.004)
```

**Comment by pjavia-iterativescopes** [comment]
Did you solve the above problem ?

**Comment by dollysapra** [comment]
Yes, i added this little code to my script to make it work. Somewhere in my conversion from onnx to caffe2 model, operators missed CUDA device option. 


 ```
device_option = core.DeviceOption(caffe2_pb2.CUDA, 0)
 
 with core.DeviceScope(device_option):
        input_data,label = AddInput(train_model,training_net_batch_size,training_lmdb_path)
  
for op in train_model.param_init_net.Proto().op :
        op.device_option.CopyFrom(device_option)

   for op in train_model.net.Proto().op :
        op.device_option.CopyFrom(device_option)
        op.engine = "CUDNN"
               
    with core.DeviceScope(device_option): 
        xent = train_model.LabelCrossEntropy(['softmax_output', label ], 'xent')
        loss = train_model.AveragedLoss(xent, "loss")
        train_model.AddGradientOperators([loss])   
        optimizer.build_sgd(train_model,base_learning_rate= lr,
                policy="fixed",momentum=0.9,
                weight_decay=w_decay)
         
    workspace.RunNetOnce(train_model.param_init_net)
    workspace.CreateNet(train_model.net, overwrite=True)

    for i in range(training_iters):
        workspace.RunNet(train_model.net)

```

Hope it helps.
